### ISSUE TITLE
Support Kserve

### ISSUE THREAD

**Issue by tenzen-y** [kubernetes-sigs/kueue#1603]
**What would you like to be added**:
I would like to support the serverless ML Inference tool, [Kserve](https://kserve.github.io/website/latest/).

**Why is this needed**:
In the hybrid workload (which means training jobs and inference servers and so on) cluster, users often want to manage all cluster capacities by the kueue's flavorQuotas. So, as the first step to support the inference server, supporting Kserve in kueue is nice to have.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.
We will probably implement `suspend` semantics on the Kserve side.
Additionally, we need to move #77 forward together to support the inference server's autoscaling semantics.

**Comment by alculquicondor** [comment]
I think I talked about this with @astefanutti 
Also cc @mwielgus @ahg-g

**Comment by ahg-g** [comment]
How do you envision that working? Can you list a couple of CUJs?

**Comment by kerthcet** [comment]
Online inference service is somehow latency sensitive, scalability is highly required, reclaim/preempt the kserve managed services looks not right. I guess Kserve is not that good at offline inference, which in my mind maybe helpful. cc@terrytangyuan

**Comment by lizzzcai** [comment]
I would like to see possible support for this as I am looking for a unified way of managing resources for both model training and serving, as in our case they are running in the same cluster. And how it can integrate with the recent `MultiKueue` feature to schedule workload to clusters with available GPU (sometimes there is a shortage of GPU in certain regions). As KServe deployment has min and max replicas, it should be scheduled to cluster that can meet the max replicas.

**Comment by tenzen-y** [comment]
> How do you envision that working? Can you list a couple of CUJs?

I imagined that the similar approach as RayCluster.

So, I would like to add Suspend field to InferenceService resource.

**Comment by tenzen-y** [comment]
> Online inference service is somehow latency sensitive, scalability is highly required, reclaim/preempt the kserve managed services looks not right. I guess Kserve is not that good at offline inference, which in my mind maybe helpful. cc @terrytangyuan

@kerthcet I believe that lending limit would allow us to guarantee capacities for latency sensitive Workloads.

**Comment by tenzen-y** [comment]
> I would like to see possible support for this as I am looking for a unified way of managing resources for both model training and serving and Kueue looks like it has this capability. In our case, both training and serving are running in the same cluster. And how it can integrate with the recent MultiKueue feature to schedule workload to clusters with available GPU (sometimes there is a shortage of GPU in certain regions). As KServe deployment has min and max replicas, it should be scheduled to cluster that can meet the max replicas.

Yes, that's right. Actually, I also deploy Job and Inference Server into a single cluster.

**Comment by tenzen-y** [comment]
Let me try to design this integrations.

/assign

**Comment by terrytangyuan** [comment]
Thanks! Great to see this. Looking forward to your proposal. @tenzen-y

**Comment by tenzen-y** [comment]
> Thanks! Great to see this. Looking forward to your proposal. @tenzen-y

I will create a dedicated issue later in Kserve side as well.

**Comment by tenzen-y** [comment]
I do not have enough bandwidth.
So, feel free to take this if anyone is interested in this feature.

/unasign

**Comment by tenzen-y** [comment]
> FYI: we have similar requirements for serving projects but mostly for resource fungibility rather quota managements(although related as a whole).

Yeah, I fully agree with you. Actually, we wanted to provide the resource fungibility in the mixed type workload clusters.
For example, I want to allocate the free resources to Jobs when a few serving requests come. After serving requests go spike, we want to preempt Jobs and allocate resources to Serving for auto scaling.

But, I found that we can fulfill this motivation by using Kueue Pod integration with a little bit of Knative tuning.
So, maybe before we start natively supporting, we may want to add documentations for the way.

I will try to add Documentation on how to perform Kserve ISVC with Kueue.

**Comment by terrytangyuan** [comment]
@tenzen-y It would be great if you could share some docs and examples on that!

**Comment by tenzen-y** [comment]
> @tenzen-y It would be great if you could share some docs and examples on that!

Sure, I will share it in the Kueue documentation.

**Comment by kannon92** [comment]
> [@tenzen-y](https://github.com/tenzen-y) It would be great if you could share some docs and examples on that!

We are looking to add a section in the docs to show using pod integration with external frameworks.

https://github.com/kubernetes-sigs/kueue/pull/3898

is an example of using Pod Integration with TektonPipelines.

@varshaprasad96 has done some exploring on how to use kubeflow notebooks with the statefulset integration.

**Comment by varshaprasad96** [comment]
> @varshaprasad96 has done some exploring on how to use kubeflow notebooks with the statefulset integration.

I'm doing a few tests with the latest version of Kueue for NB integration. Will open an upstream PR to add docs soon.

**Comment by tenzen-y** [comment]
> In case it helps - I had created this document (https://docs.google.com/document/d/1gJ-LbOG1FWVNLP-M5Wc9p5h2pNTagUzdw2hsII6BFDM/edit?tab=t.0) that goes through the process of using Kueue and integrating KServe through deployment integration. This was done on OpenShift cluster - but it is all open source and can be easily replicated on vanilla K8s too. 
> 
> If someone wants to use this doc, and convert it into Kueue docs, please feel free to make use of it :) 

Thank you for sharing it. I have serverless mode documentation, internally. We might be able to collaborate to add Kserve support documentation in Kueue.

**Comment by doctorpangloss** [comment]
Is it already possible to achieve Kueue semantics within Kserve? I.e., to limit the resource usage of "slow" requests to prevent exhausting all resources, allowing "fast" requests to continue to be executed, even when they compete for the same resources? This seems to be the crux of the value prop.

**Comment by tenzen-y** [comment]
> Is it already possible to achieve Kueue semantics within Kserve? I.e., to limit the resource usage of "slow" requests to prevent exhausting all resources, allowing "fast" requests to continue to be executed, even when they compete for the same resources? This seems to be the crux of the value prop.

I might not be able to understand it. What do "slow" requests and "fast" requests mean? Do you indicate replacement like Revision update or auto scaling?

**Comment by doctorpangloss** [comment]
My goal is to **reduce average delay of "fast" requests in exchange for longer delays of "slow" requests** when there are limited resources to serve requests, a very common scenario for ML inferencing tasks and fixed GPU quotas. Maybe even the **most** common production issue.

This is a completely orthodox, subjective design choice in the service of user experience that absolutely everyone makes. Developers using Kserve for semi-interactive function calling where the consumer is a real human being, such as with chat and art generative AI tasks: 100% of them will desire this user experience. Why? Because some requests are so much faster than others, that making the longer duration requests wait a little longer is in relative terms a small part of their total execution time. **But if there is only one queue where both long duration and short duration requests are processed, and resources are constrained, over time, short duration requests will be delayed as long as long duration requests take to execute, which means in relative terms, "short" requests become dominating by waiting for queue space.** So there's a simple solution and it's done with Kueue: concurrency limits. Let me show you **why** this works:

Let's say I have two categories of API requests: one that takes duration ~ N(24s, 2s) to complete ("short") and another that takes duration ~ N(15m, 7m) to complete ("long"). Both require exclusive access to a "GPU" resource. I have 8 GPUs.

Suppose users make one or the other request uniformly randomly (i.e., long ~ Uniform[0,1] < 0.5). Short requests tend to arrive in batches of 1-8 (pick a discrete probability dist. with a mean of 4) and long requests tend to arrive in batches of 1 OR 4. The requests tend to arrive once every 15 seconds. 

Under this arrival process things are unstable. We'll say that requests only arrive for the first 30m, and then let's simulate things for 2h.

We're going to measure the average sojourn time (wait + processing time) of the two categories of requests under two scenarios: one is that there is a single FIFO queue for all GPUs, and another is that there are two queues:

 - Long requests are limited to using at most 4 GPUs at a time. If a long request would use a 5th+   - GPU, it will wait in this "Long queue" instead.
 - Short requests can use as many GPUs as needed. This is the "short queue"
 - When a GPU that could service a long request is freed, and there is both a long and a short request waiting to use it, use FIFO (oldest time of arrival) to decide which queue to pop from for the GPU.

![Image](https://github.com/user-attachments/assets/8b8ba8d8-96e0-40dd-8197-39c97c47dcfa)

![Image](https://github.com/user-attachments/assets/0df85f00-ae14-49ac-95f7-34f3fa0cc7ad)

| Scenario | Mean sojourn at 3 h (short) | Mean sojourn at 3 h (long) | What the curves reveal |
|----------|----------------------------|----------------------------|------------------------|
| **Single FIFO queue (8 GPUs)** | **≈ 4 600 s ≈ 77 min** | **≈ 5 100 s ≈ 85 min** | Once arrivals stop (t = 30 min) backlog drains steadily, but both classes finish in roughly the same hour-plus window. |
| **Dual queue (long ≤ 4 GPUs)** | **≈ 55 s** | **≈ 5 300 s ≈ 89 min** | Short jobs stay almost interactive (< 1 min total time) because they never wait behind long ones; long jobs complete in about the same window as FIFO but without congesting the short path. |

So do you see how I added about 200s of processing time for "long" requests in exchange to only slow down "short" requests by 30s? This is a real world example. Everybody wants to do this in Kserve.

**Comment by tenzen-y** [comment]
Thank you for summarizing that. I can understand what you want to say.
I think we have 2 solutions, 1 is handling as a Kueue problem (Pod scheduling), 2 is handling as a traffic scheduling problem.

For the Pod scheduling problem,

I think 2 priorities might resolve your problem, which is priority-long and priority-short.
The long-isvc is queued with priority-long, and the short isvc is queued with priority-short which allows us to prioritize the short request by preemption.

If you are dispatching the training workloads into the same cluster, you might want to prepare dedicated cq for inference.

For the traffic scheduling problem,

I think you might introduce custom filter by envoy or introducing gateway-api-inference-extensions (GIE). However, current Kserve does not support GIE as you can see https://github.com/kserve/kserve/issues/4227
When you consider this problem as a traffic scheduling, both short and long traffic are accommodated by a single isvc, and a single cq.
